runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java

```diff
@@ -0,0 +1,317 @@
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A concrete implementation of {@link StreamingMap} that is capable of storing data
+ * off-heap. The location for the off-heap data is provided by {@link File} dataPath,
+ * while the resource policy (memory and storage limits) are defined in {@link Options}.
+ *
+ * @param <K> key type
+ * @param <V> value type
+ */
+@Slf4j
+public class PersistedStreamingMap<K, V> implements StreamingMap<K, V> {
+
+    private final StreamingMap<K, V> optimisticMap = new StreamingMapDecorator<>();
+    private final AtomicInteger dataSetSize = new AtomicInteger();
+    private final CorfuRuntime corfuRuntime;
+    private final ISerializer serializer;
+    private final RocksDB rocksDb;
+
+    public PersistedStreamingMap(@NonNull Path dataPath,
+                                 @NonNull Options options,
+                                 @NonNull ISerializer serializer,
+                                 @NonNull CorfuRuntime corfuRuntime) {
+        try {
+            FileUtils.deleteDirectory(dataPath.toFile());
+            this.rocksDb = RocksDB.open(options, dataPath.toFile().getAbsolutePath());
+        } catch (RocksDBException | IOException e) {
+            throw new UnrecoverableCorfuError(e);
+        }
+        this.serializer = serializer;
+        this.corfuRuntime = corfuRuntime;
+    }
+
+    /**
+     * A Java compatible {@link RocksIterator} implementation
+     */
+    public class RocksDbIterator implements Iterator<Entry<K, V>> {
+        private RocksIterator iterator;
+        private Entry<K, V> current;
+        private Entry<K, V> next;
+
+        RocksDbIterator(RocksIterator iterator) {
+            this.iterator = iterator;
+        }
+
+        /**
+         * Ensure that this iterator is operating under the correct assumptions.
+         */
+        private void checkInvariants() {
+            // RocksDB does not support multi-threaded access.
+            // If the iterator was created by some thread, it also has to be consumed by it.
+            if (!iterator.isOwningHandle()) {
+                throw new IllegalStateException("Detected multi-threaded access to this iterator.");
+            }
+
+            try {
+                iterator.status();
+            } catch (RocksDBException e) {
+                throw new UnrecoverableCorfuError(
+                        "There was an error reading the persisted map.", e);
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean hasNext() {
+            // If we have the next element pipelined, go ahead and return true.
+            if (next != null) {
+                return true;
+            }
+
+            // If the iterator is valid, this means that the next entry exists.
+            checkInvariants();
```

Comment:
Can you omit `iterator.status` from every call? its a native call (i.e an extra sys call)

Reply:
The documentation says to make sure that the iterator is valid before calling `next()` on it, so I am not sure if I have a choice.